### DISCUSSION TITLE
Why does the Arg decorator require argument name?

### DISCUSSION THREAD

**Discussion by capaj** [MichalLytek/type-graphql#990] [Q&A]
Currently you have to define arguments like this:
```ts
    @Arg('name', () => String) name: string,
```

why isn't the API like this instead?
```ts
    @Arg(() => String) name: string,
```

You can always get the name of the decorated argument-it's the second parameter in the decorator function. Why does typegraphql insist on me writing the name twice when it can get the name from it?

Arg decorator is implemented like this in decapi for example: https://github.com/capaj/decapi/blob/946c15853c7b8c2846c7fa12fb7bb43f39307299/src/domains/arg/ArgDecorators.ts#L6

**Comment by MichalLytek** [comment]
> You can always get the name of the decorated argument-it's the second parameter in the decorator function. 

No you can't 😅 

```ts
declare type ParameterDecorator = (target: Object, propertyKey: string | symbol, parameterIndex: number) => void;
```

**Reply by capaj** [reply]
You can't? I don't follow.
Even typescript docs define it as such:
https://www.typescriptlang.org/docs/handbook/decorators.html#parameter-decorators

![image](https://user-images.githubusercontent.com/1305378/129967524-b994e450-d3a7-4515-9759-9d84266f1471.png)

**Reply by capaj** [reply]
```
propertyKey
```
is the name of the decorated argument right there in the type you have posted. `propertyKey` is not the best name for sure. `paramName` would be much better IMHO

EDIT: nope, I was wrong

**Reply by MichalLytek** [reply]
No, `propertyKey` is the name of the method, which parameter we decorate 😉 Having only class and parameter name, we wouldn't know which method does this parameter belongs to.

Why do you think I haven't checked that by 4 years? 😛

**Reply by MichalLytek** [reply]
Closing this thread, I'm just wasting my time 🙄 

![image](https://user-images.githubusercontent.com/10618781/129970786-72727eba-8e8f-4972-be44-5bb241c888ed.png)

**Reply by capaj** [reply]
Oh very sorry! I got this confused with the method name. This was actually implemented in type-graphql by @prismake and I always taken it for granted.

Now that I looked up the implementation it is done with this regex:
https://github.com/capaj/decapi/blob/master/src/services/utils/getParameterNames.ts
plus you get the position, so you can infer the name of the param.
It's a bit hacky, but it works good on all the projects that use decapi. I've never encountered a function where the function would fail to find those param names.

**Reply by capaj** [reply]
Would you be open to a PR which would make the `name` param for `Arg()` optional?
In the case when it is not supplied, it would use the regex to find the name of the argument.

**Reply by MichalLytek** [reply]
No, regex is hacky. vNext gonna solve this by a compiler plugin, along with other duplications.
